driver-core/src/main/com/mongodb/internal/operation/OperationHelper.java

```diff
@@ -462,7 +462,11 @@ static boolean canRetryRead(final ServerDescription serverDescription, final Con
         public void onResult(final T result, final Throwable t) {
             for (ReferenceCounted cur : referenceCounted) {
                 if (cur != null) {
-                    cur.release();
+                    try {
```

Comment:
I can't say I like the looks of this. :)

Can you explain why in some cases a reference is not retained, and why this is the best way to address it?

Reply:
The issue here is we never return the result of the callback and blindly throw an exception,  breaking the callback based approach.

The error in [JAVA-3616](https://jira.mongodb.org/browse/JAVA-3616) looks to be from: 

```
FAILURE: com.mongodb.MongoSocketClosedException: Can not read from a closed socket (com.mongodb.MongoSocketClosedException)
--
com.mongodb.MongoSocketClosedException: Can not read from a closed socket
```
But that doesn't get propagated back because an IllegalStateException is thrown, leaving the callback dangling and never returning to the user.